OOP版ボウリング課題の提出物

yocaji opened the pull request:
オブジェクト指向版ボウリング課題の提出物になります。お手数ですがご確認よろしくお願いいたします🙇‍♂️

### 疑問点

- Frameクラスに比べてGameクラスが色々とやりすぎているように思いますが、適切な切り出し方が思いつかず今のような状態になっています

Commit:
OOP版ボウリング課題新規作成

Commit:
require_relative "frame"のタイミングを修正

Commit:
不要なgemの記述を削除

torinoko commented on 08.ls_object/frame.rb:
```diff
@@ -0,0 +1,17 @@
+    @shots = shots
+  end
+
+  attr_reader :shots
```
[提案]
なんとなくですが `attr_*` 系はメソッドよりも上に書くことが多い気がします 👀

yocaji replied:
`attr_`は`def initialize`より前の方が自然ということ、承知しました！

torinoko commented on 08.ls_object/frame.rb:
```diff
@@ -0,0 +1,17 @@
+  attr_reader :shots
+
+  def spare?
+    @shots.size == 2 && @shots.sum == 10 ? true : false
```
[提案]
Ruby ではメソッド内の最後の値が戻り値として返るので、明示的に true / false を返さなくても大丈夫ですよ〜 💡 
今回のケースですと `@shots.size == 2 && @shots.sum == 10` の結果が戻り値になります！

yocaji replied:
おっしゃる通りです🙏
修正しました。

torinoko commented on 08.ls_object/frame.rb:
```diff
@@ -0,0 +1,17 @@
+  end
+
+  def strike?
+    @shots.size == 1
```
[気持ち]
ストライクだと確かに size は 1 になりますね、なるほど〜 💡

torinoko commented on 08.ls_object/game.rb:
```diff
@@ -0,0 +1,44 @@
+    @score = score
+  end
+
+  attr_reader :score
```
[提案]
ここの `attr_*` も `#initialize` の上にある方が一般的かな？と思います（もし何か参考にされてこう書いているようでしたらごめんなさい 🙏 ）

torinoko commented on 08.ls_object/game.rb:
```diff
@@ -0,0 +1,44 @@
+
+  private
+
+  def convert(score)
```
[提案]
`#convert` もよい名前ですが、もう少し具体的でもいいかもしれませんね。何を何に変換するのか、を表せるメソッド名だとよりよいかなと思いました。

yocaji replied:
`convert_to_frames`に修正してみました。

torinoko replied:
@yocajii 
より具体的な名前になっていいですね〜！ ✨

torinoko commented on 08.ls_object/game.rb:
```diff
@@ -0,0 +1,44 @@
+  end
+
+  def point_of_strike
+    @frames.map.with_index do |f, i|
+      @frames[i + 1..i + 2].map(&:shots).flatten.first(2) if f.strike?
+    end.flatten.compact.sum
```
[気持ち]
`@frames[i + 1..i + 2].map(&:shots).flatten.first(2)` のところ、ちょっと複雑ですね。パッと見て何を処理しているのかわかりづらいなーと思いました。
Game クラスのメソッドが、Frame オブジェクトの内容を少し知り過ぎなのかもしれませんね。Frame オブジェクトをこねこねして利用しているので、この辺り Frame クラスなどに移動できるかも？という気がしました。

yocaji replied:
ボーナスポイントの計算をFrameクラスのほうで行うようにしてみました。

torinoko replied:
@yocajii 
Frame クラスの知識が Frame クラスに移行することで、Game クラス側の複雑さがだいぶ減りましたね！

torinoko commented on spec/08.ls_object/frame_spec.rb:
```diff
@@ -0,0 +1,20 @@
+  let(:spare) { build :frame, shots: [4, 6] }
+  let(:strike) { build :frame, shots: [10] }
+
+  example 'スペアの判定' do
+    expect(basic.spare?).to be false
+    expect(spare.spare?).to be true
+  end
```
[気持ち]
メソッドごとの判定は以下のように `describe` で囲んで、条件ごとにテストケースを書く方が RSpec らしい気がします。

```
describe '#spare?` do
  it 'スペアのときにtrueが返ること' do
    expect(spare.spare?).to be true
  end
end
```

torinoko replied:
先頭で定義している `let` の値も、必要なところで定義できるとよさそうですね（たとえば strike の値は spare? の判定で使わないですよね）

yocaji replied:
ご教示ありがとうございます✨
メソッドごとに`discribe`で分けることと、条件ごとにテストケースを書くようにすること、また`let`は必要なところで定義すること、承知しました！
`let`については、`Frame`クラスのインスタンスメソッドが1つだけになったことで都度Rubyのオブジェクトを作った方が分かりやすいように思えてきたため、ひとまず修正後のような形になっています。

torinoko replied:
@yocajii 
よいですね！
テストの値はベタ書きでよいと思います〜！
`let` などは必要に応じて使う形でよいと思うので、今回はこの形がよさそうですね。

torinoko reviewed (commented):
@yocajii 
コメントを書きました！
基本的にはよさそうに思います。

Game でいろいろやっているのはおっしゃるとおりかもしれませんね 💡 
処理を見ると、Game クラスが Frame のオブジェクトのことを詳しく知っている状態で、Game が Frame のことを少し知り過ぎな印象も受けました。
Game がもっとかんたんに Frame を扱えるといいかもしれませんね。

Commit:
レビューでいただいたコメントに基づいて修正

- ボーナスポイントの計算をFrameクラスで行うようにした
- frame_specをメソッド単位でグループ化した上で期待するケースごとに分離した
- FactoryBotを使用せずPOROでテストデータを作成するよう変更した

Commit:
利用しなくなったFactoryBotを削除した

torinoko reviewed (approved):
@yocajii 
よいと思います！ 🎖️ ✨ 
コード全体の見通しもよいですね。
不要になったライブラリとコードのお掃除もしっかりできていてばっちりです〜！
